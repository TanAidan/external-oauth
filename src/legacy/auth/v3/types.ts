import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
