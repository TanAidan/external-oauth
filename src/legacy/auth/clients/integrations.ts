// import { SetupDetailsNotFound } from '../errors'
import { TIntegrationConfig, EAuthType } from '../v3/types'
import '../../../../integrations'
import Knex from 'knex'

interface ICommonOAuthPayload {
  serviceName: string
  userId: string
  updatedAt: number
  setupId: string
  scopes?: any
  tokenResponseJSON?: string
  callbackParamsJSON?: string
  connectParams?: any
}

export type TOAuth1Payload = ICommonOAuthPayload & {
  accessToken: string
  tokenSecret: string
  consumerKey?: string
  consumerSecret?: string
  expiresIn: number
}

export type TOAuth2Payload = ICommonOAuthPayload & {
  accessToken: string
  clientId?: string
  clientSecret?: string
  refreshToken?: string
  expiresIn?: number
  idToken?: string
  idTokenJwt?: any
}

interface IAuthParams {
  buid: string
  authId: string
}

export type TOAuthPayload = TOAuth1Payload | TOAuth2Payload

export interface IAuthResult {
  id: string
  buid: string
  auth_id: string
  payload: {
    accessToken: string
    refreshToken: string
    idToken: string
    expiresIn: any
    scopes: string[]
    tokenResponseJSON: string
    updatedAt: any
    idTokenJwt: any
    connectParams?: any
    callbackParamsJSON: any
  }
}

export const updateAuth = async ({
  buid,
  authId,
  setupId,
  payload,
  store
}: IAuthParams & { payload: TOAuthPayload } & { store: any; setupId: string }) => {
  await store('authentications').insert({ buid, auth_id: authId, setup_id: setupId, payload })
}

export const getAuth = async <IAuthResult>({ buid, authId, store }: IAuthParams & { store: any }) => {
  console.log('[getAuth] buid/authId', buid, authId)

  return (await store('authentications')
    .where({ buid, auth_id: authId })
    .first()) as IAuthResult
}

export const getConfig = async ({ buid }: { buid: string }) => {
  let item = {} as any
  try {
    item = require(`../../../../integrations/${buid}.json`)
  } catch (err) {
    if (err.code === 'MODULE_NOT_FOUND') {
      return false
    }
  }

  const configItem = item.auth as TIntegrationConfig
  if (configItem.authType) {
    configItem.authType = configItem.authType.toUpperCase() as EAuthType
  }

  return { ...configItem, authConfig: configItem, requestConfig: item.request }
}

export const saveSetupDetails = async ({
  buid,
  setupId,
  credentials,
  scopes,
  store
}: {
  buid: string
  setupId: string
  store: Knex
  credentials: any
  scopes: string[] | undefined
}) => {
  await store('configurations').insert({ buid, credentials, scopes, setup_id: setupId })
}

export const getSetupDetails = async ({
  buid,
  setupId,
  store
}: {
  buid: string
  setupId: string | undefined
  store: Knex
}) => {
  if (setupId) {
    return await store('configurations')
      .where({ buid, setup_id: setupId })
      .first()
  }

  return await store('configurations')
    .where({ buid })
    .orderBy('updated_at', 'desc')
    .first()
}
