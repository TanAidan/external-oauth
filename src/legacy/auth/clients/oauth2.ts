import simpleOauth2, { Token } from 'simple-oauth2'
import Wreck from 'wreck'
import http from 'http'
import { URL } from 'url'

import { AuthenticationFailed } from '../v3/errors'
import Boom from 'boom'
import { OAuthTokenResponse } from '../v3/types'
// import { inspectAccessToken } from './openid-connect'

const headers = { 'User-Agent': 'Pizzly' }

const createClientForRedirect = ({ authorizationURL, clientId }: RedirectClientParams) => {
  const url = new URL(authorizationURL)

  return simpleOauth2.create({
    client: { id: clientId, secret: '' },
    auth: {
      tokenHost: url.origin,
      authorizeHost: url.origin,
      authorizePath: url.pathname
    },
    http: { headers }
  })
}

const createClientForToken = ({
  authorizationMethod,
  bodyFormat,
  clientId,
  clientSecret,
  tokenURL
}: TokenClientParams) => {
  const url = new URL(tokenURL)

  return simpleOauth2.create({
    client: { id: clientId, secret: clientSecret },
    auth: { tokenHost: url.origin, tokenPath: url.pathname },
    http: { headers, events: true },
    options: {
      authorizationMethod: authorizationMethod || AuthorizationMethod.Body,
      bodyFormat: bodyFormat || BodyFormat.Form
    }
  })
}

const translateError = (e: any) => {
  if (e.data && e.data.isResponseError) {
    const {
      payload,
      res: { statusCode }
    } = e.data

    const response = Buffer.isBuffer(payload) ? payload.toString() : payload
    return new AuthenticationFailed({ statusCode, response })
  }

  if (e.isBoom) {
    return new AuthenticationFailed({ message: e.message })
  }

  return e
}

const translateNoTokenResponse = (response: any) => {
  // The simple-oauth2 library blindly does an Object.assign with the response.
  // When the response was a string, it results in an object with string index
  // keys and numeric character code values!
  const hasIntegerKeys = Object.keys(response)
    .map(Number)
    .every(Number.isInteger)
  const values = Object.values(response)

  if (hasIntegerKeys && values.every(Number.isInteger)) {
    return String.fromCharCode(...(values as any))
  }

  return response
}

const getExpiresIn = async (params: TokenClientParams, token: simpleOauth2.Token) => {
  // const { clientId, clientSecret, tokenURL } = params

  // Zoho doesn't follow the spec and returns `expires_in` in ms.
  // They provide the usual value in `expires_in_sec` instead
  const expiresIn = token.expires_in_sec || token.expires_in

  // if (!expiresIn) {
  //   const metadata = await inspectAccessToken({ clientId, clientSecret, tokenURL, accessToken: token.access_token })

  //   if (metadata && metadata.exp) {
  //     const issuedAt = metadata.iat || Math.trunc(Date.now() / 1000)
  //     return metadata.exp - issuedAt
  //   }
  // }

  return expiresIn
}

const wrapTokenOperation = async (
  client: simpleOauth2.OAuthClient,
  params: TokenClientParams,
  body: () => Promise<Token>
) => {
  // The simple-oauth2 library has no supported way of accessing the
  // request/response so we must reach into it's guts and hook the HTTP client
  // that it uses!
  const wreck = (client.authorizationCode as any).client.client as typeof Wreck

  let headers
  wreck.events!.once('response', (_err: Boom | undefined, { res }: WreckResponseDetails) => {
    if (res) {
      headers = res.headers
    }
  })

  let token
  try {
    token = await body()
  } catch (e) {
    throw translateError(e)
  }

  if (!token.access_token) {
    throw new AuthenticationFailed({
      message: 'No access token returned',
      response: translateNoTokenResponse(token)
    })
  }

  return {
    accessToken: token.access_token,
    expiresIn: await getExpiresIn(params, token),
    idToken: token.id_token,
    refreshToken: token.refresh_token,
    decodedResponse: {
      headers,
      body: token
    }
  } as TokenResult
}

const buildScope = (scopes: string[]) => scopes.join(' ')

export const getCodeRedirectURL = (params: RedirectParams) => {
  const { authorizationParams, callbackURL, scope, state } = params
  const client = createClientForRedirect(params)

  return client.authorizationCode.authorizeURL({
    ...authorizationParams,
    state,
    redirect_uri: callbackURL,
    scope: buildScope(scope)
  })
}

export const getTokenWithCode = async (params: CodeParams) => {
  const { callbackURL, code, tokenParams } = params
  const client = createClientForToken(params)

  return wrapTokenOperation(client, params, () =>
    client.authorizationCode.getToken({
      ...tokenParams,
      code,
      redirect_uri: callbackURL
    })
  )
}

export const getTokenWithRefreshToken = async (params: RefreshParams) => {
  const { idToken, refreshToken } = params
  const client = createClientForToken(params)
  const clientToken = client.accessToken.create({ refresh_token: refreshToken })

  const response = await wrapTokenOperation(client, params, async () => {
    return (await clientToken.refresh()).token
  })

  response.refreshToken = response.refreshToken || refreshToken
  response.idToken = response.idToken || idToken

  return response
}

export const getTokenWithClientCredentials = async (params: GetTokenClientCredsParams) => {
  const { scope } = params
  const client = createClientForToken(params)

  return wrapTokenOperation(client, params, async () => {
    return client.clientCredentials.getToken({ scope: buildScope(scope) })
  })
}

interface RedirectClientParams {
  authorizationURL: string
  clientId: string
}

interface RedirectParams extends RedirectClientParams {
  authorizationParams: any
  callbackURL: string
  scope: string[]
  state: string
}

export enum AuthorizationMethod {
  Body = 'body',
  Header = 'header'
}

export enum BodyFormat {
  Form = 'form',
  JSON = 'json'
}

interface TokenClientParams {
  authorizationMethod?: AuthorizationMethod
  bodyFormat?: BodyFormat
  clientId: string
  clientSecret: string
  tokenURL: string
}

interface CodeParams extends TokenClientParams {
  callbackURL: string
  code: any // string
  tokenParams: any
}

interface RefreshParams extends TokenClientParams {
  idToken?: string
  refreshToken: string
}

interface GetTokenClientCredsParams extends TokenClientParams {
  scope: string[]
}

export interface TokenResult {
  accessToken: string
  expiresIn?: number
  idToken?: string
  refreshToken?: string
  decodedResponse: OAuthTokenResponse
}

interface WreckResponseDetails {
  req: http.ClientRequest
  res?: http.IncomingMessage
  start: number
  url: URL
}
