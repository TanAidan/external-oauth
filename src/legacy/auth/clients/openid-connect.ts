import { URL } from 'url'
import axios from 'axios'
import querystring from 'querystring'

const inspectHeaders = {
  'Content-Type': 'application/x-www-form-urlencoded',
  'User-Agent': 'Bearer.sh'
}

const fetchConfiguration = async (issuerIdentifierURL: string): Promise<OpenIdConfiguration | undefined> => {
  try {
    return (await axios.get(`${issuerIdentifierURL}/.well-known/openid-configuration`)).data
  } catch (e) {
    console.error(e)
  }
}

const inspectToken = async ({
  clientId,
  clientSecret,
  introspectionEndpoint,
  token,
  tokenType
}: InspectTokenParams): Promise<TokenMetadata | undefined> => {
  const body = querystring.stringify({ token, token_type_hint: tokenType })
  const options = { auth: { username: clientId, password: clientSecret }, headers: inspectHeaders }

  try {
    return (await axios.post(introspectionEndpoint, body, options)).data
  } catch (e) {
    console.error(e)
  }
}

export const inspectAccessToken = async ({
  accessToken,
  clientId,
  clientSecret,
  tokenURL
}: InspectAccessTokenParams) => {
  const parsedTokenURL = new URL(tokenURL)
  const config = await fetchConfiguration(parsedTokenURL.origin)

  if (config && config.introspection_endpoint) {
    return inspectToken({
      clientId,
      clientSecret,
      introspectionEndpoint: config.introspection_endpoint,
      token: accessToken,
      tokenType: TokenType.AccessToken
    })
  }
}

interface InspectAccessTokenParams {
  accessToken: string
  clientId: string
  clientSecret: string
  tokenURL: string
}

interface InspectTokenParams {
  clientId: string
  clientSecret: string
  introspectionEndpoint: string
  token: string
  tokenType: TokenType
}

interface OpenIdConfiguration {
  introspection_endpoint?: string
}

enum TokenType {
  AccessToken = 'access_token'
}

export interface TokenMetadata {
  exp?: number
  iat?: number
}
