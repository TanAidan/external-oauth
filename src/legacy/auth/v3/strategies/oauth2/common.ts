import jwt from 'jsonwebtoken'

import { TokenResult } from '../../../clients/oauth2'
import { IOAuth2Credentials } from '../../types'

export const NO_VALUE = 'non'

export const responseToCredentials = ({ accessToken, refreshToken, idToken, expiresIn }: TokenResult) => {
  const credentials: IOAuth2Credentials = {
    accessToken,
    refreshToken: refreshToken || NO_VALUE,
    idToken: idToken || NO_VALUE,
    expiresIn: expiresIn || 0
  }

  if (idToken) {
    credentials.idTokenJwt = getIdTokenJwt(idToken)
  }

  return credentials
}

export const getIdTokenJwt = (idToken?: string) => {
  if (!idToken) {
    return undefined
  }

  try {
    return jwt.decode(idToken) || undefined
  } catch (e) {
    return undefined
  }
}
