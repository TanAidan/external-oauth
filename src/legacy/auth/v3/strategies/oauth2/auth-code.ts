import { NextFunction, Response } from 'express'

import { TAuthenticateRequest } from '../../types'
import { getCodeRedirectURL, getTokenWithCode } from '../../../clients/oauth2'
import { asyncMiddleware } from '../../../../errorHandler'
import { responseToCredentials } from './common'
import { AuthenticationFailed } from '../../errors'

export const authenticate = asyncMiddleware(async (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  const callbackURL = process.env.AUTH_CALLBACK_URL || `${req.protocol}://${req.get('host')}/auth/callback`
  const {
    credentials: { clientId, clientSecret },
    scopes = []
  } = req.setupDetails
  const { code, error } = req.query
  const {
    authorizationURL,
    authorizationMethod,
    authorizationParams,
    bodyFormat,
    config,
    tokenParams,
    tokenURL
  } = req.integrationConfig

  if (error) {
    throw AuthenticationFailed.fromOAuthRequest(req, undefined)
  }

  if (code) {
    const tokenResult = await getTokenWithCode({
      authorizationMethod,
      bodyFormat,
      clientId,
      clientSecret,
      code,
      tokenParams,
      tokenURL,
      callbackURL
    })

    // console.log('tokenResult', tokenResult)
    req.credentials = responseToCredentials(tokenResult)
    // console.log('credentials', req.credentials)
    // console.log('decodedToken', tokenResult.decodedResponse)
    req.tokenResponse = tokenResult.decodedResponse

    return next()
  }

  // const { scope = [], state = 'none' } = config || {}
  const { state = 'none' } = config || {}

  const redirectURL = getCodeRedirectURL({
    authorizationParams,
    authorizationURL,
    clientId,
    state,
    scope: scopes || config?.scope || [],
    callbackURL
  })

  res.redirect(redirectURL)
})
