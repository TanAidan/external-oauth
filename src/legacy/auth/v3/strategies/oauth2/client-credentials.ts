import { NextFunction, Response } from 'express'

import { TAuthenticateRequest } from '../../types'
import { getTokenWithClientCredentials } from '../../../clients/oauth2'
import { asyncMiddleware } from '../../../../errorHandler'
import { responseToCredentials } from './common'

export const authenticate = asyncMiddleware(async (req: TAuthenticateRequest, _res: Response, next: NextFunction) => {
  const { authorizationMethod, bodyFormat, config, tokenURL } = req.integrationConfig
  const { scope = [] } = config || {}
  const { clientId, clientSecret } = req.setupDetails.credentials

  const tokenResult = await getTokenWithClientCredentials({
    authorizationMethod,
    bodyFormat,
    clientId,
    clientSecret,
    scope,
    tokenURL
  })

  req.credentials = responseToCredentials(tokenResult)
  req.tokenResponse = tokenResult.decodedResponse

  next()
})
