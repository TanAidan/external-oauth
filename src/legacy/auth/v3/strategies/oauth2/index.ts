import { NextFunction, Response } from 'express'
// import merge from 'merge'

import { TAuthenticateRequest, IFetchAuthDetailsParams, TIntegrationConfig } from '../../types'
import { /* AccessTokenExpired,*/ InvalidGrantType, InvalidAuthId } from '../../errors'
import {
  getAuth,
  IAuthResult
  // getSetupDetails,
  // updateAuthV3
} from '../../../clients/integrations'
// import { checkSetupIdConsistency } from '../setup-id-consistency'
// import { getTokenWithRefreshToken, getTokenWithClientCredentials } from '../../../clients/oauth2'
import { authenticate as authCodeAuthenticate } from './auth-code'
import { authenticate as clientCredentialsAuthenticate } from './client-credentials'
import { /*NO_VALUE, responseToCredentials,*/ getIdTokenJwt } from './common'
// import { expandAuthConfig } from '../../../../api-config/auth-config'
// import { TOAuth2Payload } from '../../../clients/integrations'

export enum GrantType {
  AuthCode = 'authorization_code',
  ClientCredentials = 'client_credentials'
}

export const CLOCK_DRIFT_MS = 60 * 1000

export const nullAuthDetails = {
  idToken: null,
  idTokenJwt: null,
  accessToken: null
}

const strategies = {
  [GrantType.AuthCode]: authCodeAuthenticate,
  [GrantType.ClientCredentials]: clientCredentialsAuthenticate
}

export const authenticate = (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  const { grantType = GrantType.AuthCode } = req.integrationConfig

  if (!Object.values(GrantType).includes(grantType)) {
    throw new InvalidGrantType(grantType)
  }

  strategies[grantType](req, res, next)
}

// const performRefresh = async ({
//   existingCredentials,
//   integrationConfig
//   // setupDetails
// }) => {
//   const { refreshToken, idToken } = existingCredentials
//   // const { clientId, clientSecret } = setupDetails
//   const { clientId, clientSecret } = { clientId: 'clientId', clientSecret: 'clientSecret' }

//   const { authorizationMethod, bodyFormat, config, grantType, refreshURL, tokenURL } = integrationConfig

//   if (refreshToken === NO_VALUE) {
//     if (grantType === GrantType.ClientCredentials) {
//       const { scope = [] } = config || {}

//       return await getTokenWithClientCredentials({
//         authorizationMethod,
//         bodyFormat,
//         clientId,
//         clientSecret,
//         scope,
//         tokenURL
//       })
//     }

//     throw new AccessTokenExpired()
//   }

//   return await getTokenWithRefreshToken({
//     ...{ clientId: 'ClientID', clientSecret: 'clientSecret' },
//     // ...setupDetails,
//     refreshToken,
//     idToken,
//     tokenURL: refreshURL || tokenURL
//   })
// }

// const refreshAndUpdateCredentials = async (
//   params: IFetchAuthDetailsParams,
//   existingCredentials: TOAuth2Payload,
//   integrationConfig: TIntegrationConfig
// ) => {
//   const { integration, authId } = params
//   const { callbackParamsJSON, connectParams = {}, setupId, tokenResponseJSON } = existingCredentials

//   // const setupDetails = await getSetupDetails({
//   //   scopedUserDataTableName,
//   //   setupId,
//   //   buid: integration.buid,
//   //   clientId: environmentIdentifier
//   // })

//   const tokenResult = await performRefresh({
//     existingCredentials,
//     // setupDetails,
//     integrationConfig: expandAuthConfig({ connectParams, authConfig: integrationConfig as any })
//   })
//   const credentials = responseToCredentials(tokenResult)

//   const previousResponse = tokenResponseJSON ? JSON.parse(tokenResponseJSON) : {}
//   const tokenResponse = merge.recursive(previousResponse, tokenResult.decodedResponse)
//   const updatedAt = Date.now()

//   const oauthPayload: TOAuth2Payload = {
//     setupId,
//     updatedAt,
//     tokenResponseJSON: JSON.stringify(tokenResponse),
//     serviceName: integration.buid,
//     userId: authId!,
//     ...credentials
//   }

//   if (callbackParamsJSON) {
//     oauthPayload.callbackParamsJSON = callbackParamsJSON
//   }

//   // await updateAuthV3({
//   //   servicesTableName,
//   //   payload: oauthPayload,
//   //   buid: integration.buid,
//   //   authId: authId!,
//   //   clientId: environmentIdentifier
//   // })

//   const callbackParams = callbackParamsJSON ? JSON.parse(callbackParamsJSON) : undefined

//   // const { clientId, clientSecret } = setupDetails
//   const { clientId, clientSecret } = { clientId: 'clientId', clientSecret: 'ClientSecret' }
//   const { accessToken, expiresIn, idToken, idTokenJwt, refreshToken } = credentials

//   return {
//     accessToken,
//     callbackParams,
//     clientId,
//     clientSecret,
//     connectParams,
//     expiresIn,
//     idToken,
//     idTokenJwt,
//     refreshToken,
//     tokenResponse,
//     updatedAt
//   }
// }

export const fetchAuthDetails = async (params: IFetchAuthDetailsParams, integrationConfig: TIntegrationConfig) => {
  const { buid, integration, authId, store, configuration } = params

  const authentication = await getAuth<IAuthResult>({
    store,
    buid: integration.buid,
    authId: authId!
  })

  if (!authentication || !authentication.payload || !authentication.payload.accessToken) {
    throw new InvalidAuthId(buid, authId!)
  }

  const {
    callbackParamsJSON,
    connectParams = {},
    expiresIn,
    updatedAt,
    accessToken,
    idToken,
    idTokenJwt,
    refreshToken,
    tokenResponseJSON
  } = authentication.payload
  const tokenResponse = tokenResponseJSON ? JSON.parse(tokenResponseJSON) : undefined
  const callbackParams = callbackParamsJSON ? JSON.parse(callbackParamsJSON) : undefined

  // checkSetupIdConsistency({ setupId, setupIdParam, setupIdFromRequest })

  const { clientId, clientSecret } = configuration.credentials
  if (!clientId || !clientSecret) {
    throw new Error('Configuration is lacking of a clientId and/or clientSecret')
  }

  // if (expiresIn) {
  //   // Sorry to be that explicit but that's something so hard to understand
  //   const expiredFromThisTime = expiresIn * 1000 + updatedAt

  //   const safeRefreshTime = expiredFromThisTime - CLOCK_DRIFT_MS

  //   const isNowLaterThanSafeRefreshTime = Date.now() > safeRefreshTime

  //   if (isNowLaterThanSafeRefreshTime) {
  //     return refreshAndUpdateCredentials(params, credentials, integrationConfig)
  //   }
  // }

  return {
    accessToken,
    callbackParams,
    connectParams,
    expiresIn,
    idToken,
    refreshToken,
    tokenResponse,
    updatedAt,
    clientId,
    clientSecret,
    idTokenJwt: idTokenJwt || getIdTokenJwt(idToken)
  }
}
