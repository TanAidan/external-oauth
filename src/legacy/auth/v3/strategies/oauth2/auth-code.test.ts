import { MiddlewareTestHarness } from '../../../../../../tests/utils'
import { authenticate } from './auth-code'
import { TAuthenticateRequest, EAuthType, TIntegrationConfig } from '../../types'
import { getCodeRedirectURL } from '../../../clients/oauth2'
import { responseToCredentials } from './common'

jest.mock('../../../../clients/oauth2')
jest.mock('./common')

const requiredIntegrationConfig = {
  authType: EAuthType.OAuth2,
  authorizationURL: 'https://example.com/auth',
  tokenURL: 'https://example.com/token',
  config: {}
}

const fullIntegrationConfig = {
  ...requiredIntegrationConfig,
  authorizationParams: {
    someAuthParam: 'test-ap'
  },
  tokenParams: {
    someTokenParam: 'test-tp'
  },
  config: {
    scope: ['test-scope1', 'test-scope2'],
    state: 'test-state'
  }
}

const setupDetails = {
  clientId: 'test-client-id',
  clientSecret: 'test-client-secret'
}

describe('authenticate', () => {
  const setup = (integrationConfig: TIntegrationConfig = fullIntegrationConfig) =>
    new MiddlewareTestHarness<TAuthenticateRequest>({
      configureRequest: req => {
        req.integrationConfig = integrationConfig
        req.setupDetails = setupDetails
      },
      testMiddleware: authenticate
    })

  describe('when connecting', () => {
    const redirectURL = 'https://example.com/auth?with_params=true'

    it('redirects to the authorization url with correct params', async () => {
      // mocked(getCodeRedirectURL)
      //   .mockReturnValueOnce(redirectURL)
      //   .mockClear()

      await setup()
        .get()
        .expect(302)
        .expect('Location', redirectURL)

      expect(getCodeRedirectURL).toMatchSnapshot()
    })
  })

  describe('when called back by the third party', () => {
    const accessToken = 'test-access-token'
    const refreshToken = 'test-refresh-token'
    const idToken = 'test-id-token'
    const expiresIn = 1234

    const tokenResult = { accessToken: 'response-access-token', decodedResponse: { body: { hey: 'there' } } }
    const credentials = { accessToken, expiresIn, idToken, refreshToken }

    beforeAll(() => {
      // mocked(getTokenWithCode).mockResolvedValue(tokenResult)
      // mocked(responseToCredentials).mockReturnValue(credentials)
    })

    beforeEach(() => {
      // mocked(responseToCredentials).mockClear()
    })

    it('exchanges the code for tokens and stores them on the request', async () => {
      const test = setup()

      await test
        .get()
        .query('code=test-code')
        .expect(200)

      expect(responseToCredentials).toHaveBeenCalledWith(tokenResult)
      expect(test.req.credentials).toEqual(credentials)
    })

    it('stores the token call response data on the request', async () => {
      const test = setup()

      await test
        .get()
        .query('code=test-code')
        .expect(200)

      expect(test.req.tokenResponse).toEqual(tokenResult.decodedResponse)
    })

    it('raises an AuthenticationFailed error if there is an error query parameter', async () => {
      const test = setup()
      await test
        .get()
        .query('error=test_code&error_description=oops')
        .expect(403)

      expect(test.err).toMatchSnapshot()
    })
  })
})
