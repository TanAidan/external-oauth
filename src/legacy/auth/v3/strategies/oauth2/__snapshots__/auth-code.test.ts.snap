// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate when called back by the third party raises an AuthenticationFailed error if there is an error query parameter 1`] = `
[AuthenticationFailed: Authentication failed:
{
  "error": "test_code",
  "error_description": "oops"
}]
`;

exports[`authenticate when connecting redirects to the authorization url with correct params 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "authorizationParams": Object {
          "someAuthParam": "test-ap",
        },
        "authorizationURL": "https://example.com/auth",
        "callbackURL": "https://example.com/v2/auth/callback",
        "clientId": "test-client-id",
        "scope": Array [
          "test-scope1",
          "test-scope2",
        ],
        "state": "test-state",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": "https://example.com/auth?with_params=true",
    },
  ],
}
`;
